pdbundler.1.md
---
title: "pdbundler(1) user manual | 0.0.6 3f754d7"
pubDate: 2024-02-15
author: "R. S. Doiel"
---

# NAME

`pdbundler`

# SYNOPSIS

`pdbundler [OPTIONS] YAML_CONFIG_FILE`

# DESCRIPTION

`pdbundler` is web service that takes a JSON object and bundles it for use with a Pandoc's web service. It requires a configuration file that maps requests to a template and options. When the `pdbundler` web service is running it accepts JSON and returns JSON suitable to POST to Pandoc web service. If an error is encounter than the response is an HTTP error response.

In the first prototype the Pandoc web service was integrated into the Newt data router. In the second prototype the Newt data router supports a generalized data pipeline. The Newt data router makes no assumptions tieing it to a template engine. This means the output from Solr or Postgres+PostgREST needs to be transformed into a new JSON expression to be usabled by Pandoc web service. `pdbundler` provides this support. It should run in the pipeline between your JSON data source and just before Pandoc web service.

`pdbundler` is a web service configured from a YAML file. It has a syntax specific to its purposes. `pdbundler` is an example of a narrowly focuses web service or micro service.

Initialization process.

1. `pdbundler` read the YAML configuration file
2. Builds request handlers mapping URL requests with a template's source
3. Start up an http service on local host listening on the specified port for requests.

When the web service is active 

- `pdbundler` receives a request in the form of a HTTP method and URL
- It checks it handlers for a match. If none are found a 404 is returned.
- With a matched request, it builds an object setting the values of `.text` and `.template` accordingly[^1]. 

### Features

- templates are read in at startup and are retained in memory bound to the request path
- the template descriptions can include Pandoc configuration objects to send to Pandoc web service
- configuration options can be set through variables in the URL Path of the request associated with the template

These features are intended to expose the capabilities of Pandoc web service.

# OPTIONS

-help
: Display this help message

-license
: Display license information

-version
: Display version information

-port
: Run service on designated port

## Pandoc Bundler's YAML

port
: (integer, defaults to 3029) the port number the service should listen on

templates
: (list of template paths and options)

### a template description

An individual template description has four properties.

`request: [METHOD] PATTERN`
: (required) This binds a request to a template. METHOD is usually either a GET or a POST. If METHOD is not defined then a POST is assumed. PATTERN is usually the path to associated with a template name. The PATTERN supports the syntax described in Go 1.22 `http` package, see <https://pkg.go.dev/net/http@master#hdr-Patterns>. Variables expressed in PATTERN are merged into the resulting JSON object. They overwrite what is set in the `.options` settings..

template
: (string, optional) This is path to the primary template used required for this request. The source if read and associated with this request signature. If template is not provided then Pandoc server will default to its internal templates

options
: (object, optional) Are used to form the root object properties in the returned JSON. This is where you would specify the Pandoc web service processing options. The options control the transformation of the text submitted[^2]. NOTE: the `.text` property is replaced by the text received by `pdbundler` and `.template` will be replaced with the resolved source read in at `pdbundler` startup.

## Example YAML configuration

This example shows six different template options. The first three apply one custom `page.tmpl` in different ways. In the last three do the same bu assume the default Pandoc template.

~~~yaml
port: 3029
templates:
  - request: "POST /custom_page"
    template: page.tmpl
    options:
      from: markdown
      to: html5
      standalone: true
      title: This is the custom template with this title
  - request: "POST /custom_page_with_title/{title}"
    template: page.tmpl
    options:
      from: markdown
      to: html5
      standalone: true
      title: This title is overwritten by the one in the request
  - request: "POST /custom_page_include"
    template: page.tmpl
    options:
      from: markdown
      to: html5
      standalone: false
  - request: "POST /default_html5"
    options:
      from: markdown
      to: html5
      standalone: true
      title: A Page using the default template
  - request: "POST /default_html5_with_title/{title}"
    options:
      from: markdown
      to: html5
      standalone: true
      title: This title is replaced by the title in the URL
  - request: "POST /default_html5_include"
    options:
      from: markdown
      to: html5
      standalone: false
~~~

[^1]: See <https://pandoc.org/MANUAL.html#running-pandoc-as-a-web-server> for basic information about Pandoc web service. Explore the website to get familiar with Pandoc and its web service.

[^2]: On <https://pandoc.org/MANUAL.html#running-pandoc-as-a-web-server> there is a link, "pandoc-server", to the current release manual page. This covers the details of using Pandoc web service including the POST JSON object setup.



ancillary/pdbundler.go
package newt

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"path"
	"strings"

	// 3rd party packages
	"gopkg.in/yaml.v3"
)

// PandocBundler models the application `pdbundler`
type PandocBundler struct {
	Port string `json:"port,omitempty" yaml:"port"`
	Templates []*PandocTemplate `json:"templates,omitempty" yaml:"templates"`
}

// PandocTemplate hold the request to template mapping for `pdbundler`
type PandocTemplate struct {
	// Pattern holds a request pattern, e.g. `POST /blog_post`
	// A request is associated with a template to be bundled into
	// an JSON object. The pattern conforms to Go 1.22 or later's
	// HTTP handler function pattern, see <https://tip.golang.org/doc/go1.22#enhanced_routing_patterns>
	Pattern string `json:"request,required" yaml:"request"`

	// Template holds a path to the primary template file for this route. Path can be relative
	// to the current working directory.
	Template string `json:"template,required" yaml:"template"`

	// Options hold the JSON object that will be resolve by `tmplbndl`. The values `.text` and `.template`
	// will be replaced by the contents specified in Bundles and received in the request.
	Options map[string]interface{} `json:"options,omitempty" yaml:"options"`

	// Debug logs more verbosely if true
	Debug bool `json:"debug,omitempty" yaml:"debug"`

	// Src holds the resolved template content
	Src []byte

	// Vars holds the names of any variables expressed in the pattern, these an be used to replace elements of
	// the output object.
	Vars []string
}

// NewPandocBundler create a new PandocBundler struct. If a filename
// is provided it reads the file and sets things up accordingly.
func NewPandocBundler(fName string, cfg *Config) (*PandocBundler, error) {
	pb := &PandocBundler{}
	if fName == "" {
		return pb, fmt.Errorf("missing configuration file")
	}
	src, err := os.ReadFile(fName)
	if err != nil {
		return nil, err
	}
	err = yaml.Unmarshal(src, &pb)
	if err != nil {
		return nil, err
	}
	if pb.Port == "" && cfg.Application != nil &&  cfg.Application.Port != 0 {
		pb.Port = fmt.Sprintf(":%d", cfg.Application.Port)
	}
	// Prefix the port number with a colon
	if ! strings.HasPrefix(pb.Port, ":") {
		pb.Port = fmt.Sprintf(":%s", pb.Port)
	}
	return pb, nil
}

// ResolvesTemplate. If not template name is available it is assumed
// you're going to use one of Pandoc default templates. If a name is
// provided then it reads the file saving the results in `.Src`
// An error is returned in a problem is encountered.
func (pt *PandocTemplate) ResolveTemplate() error {
	if pt.Template != "" {
 		src, err := os.ReadFile(pt.Template)
		if err != nil {
			return err
		}
		pt.Src = src
	}
	return nil
}

// Handler provides a HandleFunc for use with an http.ServeMux struct.
func (pt *PandocTemplate) Handler(w http.ResponseWriter, r *http.Request) {
	if pt.Debug {
		log.Printf("DEBUG .Handler(w, %s %s)", r.Method, r.URL.Path)
	}
	obj := map[string]interface{}{}
	// Copy in our options
	if pt.Options != nil {
		for k, v := range pt.Options {
			obj[k] = v
		}
	}
	// Merge in path values into options
	for _, varname := range pt.Vars {
		val := r.PathValue(varname)
		if val != "" {
			obj[varname] = val
		}
	}
	if pt.Debug {
		log.Printf("DEBUG varnames -> %+v\n", pt.Vars)
		log.Printf("DEBUG obj now -> %+v\n", obj)
	}
	// Add our resolved template source
	obj["template"] = fmt.Sprintf("%s", pt.Src)
	// Add our `.text` value from the request body
	body, err := io.ReadAll(r.Body)
	if err != nil {
		// Need to return an HTTP 400 error status. r.Body wasn't welformed
		w.WriteHeader(http.StatusBadRequest)
		w.Write([]byte(http.StatusText(http.StatusBadRequest)))
		return
	}
	body = bytes.TrimSpace(body)
	// Is the body a JSON object or some other binary object?
	switch {
		case bytes.HasPrefix(body, []byte("{")):
			// Handle case of JSON object
			m := map[string]interface{}{}
			if err := json.Unmarshal(body, &m); err != nil {
				// Need to return an HTTP 400 error status. r.Body wasn't welformed
				w.WriteHeader(http.StatusBadRequest)
				w.Write([]byte(fmt.Sprintf("%s, %s", http.StatusText(http.StatusBadRequest), err)))
				return
			}
			src, _ := yaml.Marshal(m)
			obj["text"] = fmt.Sprintf("%s", src)
		case bytes.HasPrefix(body, []byte("[")):
			// Handle case of JSON array
			a := []interface{}{}
			if err := json.Unmarshal(body, &a); err != nil {
				// Need to return an HTTP 400 error status. r.Body wasn't welformed
				w.WriteHeader(http.StatusBadRequest)
				w.Write([]byte(fmt.Sprintf("%s, %s", http.StatusText(http.StatusBadRequest), err)))
				return
			}
			src, _ := yaml.Marshal(a)
			obj["text"] = fmt.Sprintf("%s", src)
		default:
			// If we a text format then we just attach to `.text` in obj. Otherwise we need to encode it.
			contentType := http.DetectContentType(body)
			if strings.HasPrefix(contentType, "text/") {
				obj["text"] = body
			} else {
				dst := make([]byte, base64.StdEncoding.EncodedLen(len(body)))
				base64.StdEncoding.Encode(dst, body)
				obj["text"] = dst
			}
	}
	// Build a respond with a object and JSON encode it.
	src, err := json.MarshalIndent(obj, "", "    ")
	if err != nil {
		// Need to return an HTTP 400 error status. r.Body wasn't welformed
		w.WriteHeader(http.StatusUnprocessableEntity)
		// FEAT: Be nice to allow a debug hook to debug content problems before writing error ...
		w.Write([]byte(http.StatusText(http.StatusUnprocessableEntity)))
		return
	}
	w.Write(src)
}


// ResolvePath reviews the `.Request` attribute and updates the Vars using PatternKeys()
func (pt *PandocTemplate) ResolvePath() error {
	// Does the `.Request` hold a pattern or a fixed string?
	if strings.Contains(pt.Pattern, "{") {
		if ! strings.Contains(pt.Pattern, "}") {
			return fmt.Errorf("%q is malformed", pt.Pattern)
		}
		// Record our list of var names so handler can override the object being constructed from a path.
		pt.Vars = PatternKeys(pt.Pattern)
	}
	if pt.Debug {
		log.Printf("DEBUG assigning pt.Pattern -> %q\n", pt.Pattern)
		log.Printf("DEBUG vars -> %+v\n", pt.Vars)
	}
	return nil
}

func (pb *PandocBundler) ListenAndServe() error {
	mux := http.NewServeMux()
	for _, bndl := range pb.Templates {
		mux.HandleFunc(bndl.Pattern, func(w http.ResponseWriter, r *http.Request) {
			if bndl.Debug {
				log.Printf("DEBUG mux.HandleFunc(%q, bndl.Handler)", bndl.Pattern)
				log.Printf("DEBUG .vars -> %+v", bndl.Vars)
			}
			bndl.Handler(w, r)
		})
	}
	// Now create my http server
	svr := new(http.Server)
	svr.Addr = pb.Port
	svr.Handler = mux
	log.Printf("%s listening on %s", path.Base(os.Args[0]), svr.Addr)
	if err := svr.ListenAndServe(); err != nil {
		log.Fatal(err)
	}
	return nil
}

